protocols/messages.go

```diff
@@ -121,12 +124,26 @@ func (p *messagePayload) MarshalJSON() ([]byte, error) {
 	return json.Marshal(m)
 }
 
+// ErrInvalidPayload is returned when the payload has too many values
+var ErrInvalidPayload = fmt.Errorf("payload has too many values")
+
 // DeserializeMessage deserializes the passed string into a protocols.Message.
 func DeserializeMessage(s string) (Message, error) {
 	msg := jsonMessage{}
 	err := json.Unmarshal([]byte(s), &msg)
 
-	// TODO: validation
+	for _, p := range msg.Payloads {
+		numPresent := 0
+		if p.hasProposal() {
+			numPresent += 1
+		}
+		if p.hasState() {
+			numPresent += 1
+		}
+		if numPresent != 1 {
+			return Message{}, ErrInvalidPayload
+		}
+	}
```

Comment:
Is there an opportunity to DRY things out a bit by reusing the `.Type()`? method?

Reply:
Unfortunately not, since the correctness of `.Type()` depends on this validation passing.

This change is actually shifting the responsibility of validating data:
- before, data was validated every time `Type()` was called. Adding new types makes this validation awkward (imagine `hasProposal() && !hasState() && !hasPoke() && !hasRejection() && !hasAddressChange()` on every line)
- now, data is validated when the message is constructed. Messages are constructed in two ways: i) a constructor and ii) by deserializing incoming messages

Alice's wallet will create **new** messages via constructors. We can assume these messages are safe. Alice may receive an unsafe message from Eve, so we should validate when deserializing, but from that point on we can assume the messages are safe. (This is of course implicitly assuming that payloads are immutable once created.)

All in all, I think that the pattern used at the present is risky if we care about "valid" union types (ie. this Payload is **exactly one of** a SignedStatePayload and a SignedProposalPayload)

Another idea is to put a static `type` property on the payload. On a payload of type `SignedState`, external packages would not see any data other than the SignedState. I'd probably prefer that to the status quo, but would do that in a different PR. (This PR was about removing Payload from the `protocols` package API)